MINOR: Improve consumer rebalance callbacks docs

lianetm opened the pull request:
Clarify rebalance callbacks behaviour (got some questions for
onPartitionsAssigned, docs where indeed confusing about the partitions
received in params).   Reviewed all rebalance callbacks with it.

Reviewers: Bill Bejeck<bbejeck@apache.org>

lianetm committed:
docs

lianetm committed:
update

lianetm committed:
update

lianetm committed:
clean up

bbejeck commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java:
```diff
@@ -120,13 +120,31 @@ public interface ConsumerRebalanceListener {
+     * If the consumer is using the {@link GroupProtocol#CLASSIC} rebalance protocol:
+     * <ul>
+     *     <li>
+     *         In eager rebalancing, onPartitionsRevoked will be called with the full set of assigned partitions as parameter (all partitions are revoked).
```
```suggestion
     *         In eager rebalancing, onPartitionsRevoked will be called with the full set of assigned partitions as a parameter (all partitions are revoked).
```

bbejeck commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java:
```diff
@@ -120,13 +120,31 @@ public interface ConsumerRebalanceListener {
+     *         It will be called even if there are no partitions to revoke.
+     *     </li>
+     *     <li>
+     *         In cooperative rebalancing, onPartitionsRevoked will be called with the set of partitions to revoked,
```
```suggestion
     *         In cooperative rebalancing, onPartitionsRevoked will be called with the set of partitions to revoke
```

bbejeck commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java:
```diff
@@ -120,13 +120,31 @@ public interface ConsumerRebalanceListener {
+     *     </li>
+     * </ul>
+     * If the consumer is using the {@link GroupProtocol#CONSUMER} rebalance protocol, this callback will be called
+     * with the set of partitions to revoked iff the set is non-empty
```
```suggestion
     * with the set of partitions to be revoked iff the set is non-empty
```

bbejeck reviewed (approved):
Nice improvement @lianetm LGTM with a few minor comments.

lianetm committed:
Adding tests for Classic-Eager, Classic-Coop, Consumer

lianetm committed:
Address comments

lianetm committed:
More clarifications around Eager/Coop/Consumer

lianetm commented:
Thanks for the review @bbejeck ! All comments addressed. 
Also added 3 integration tests to showcase the behaviour, and added so more clarifications around it. Please take another look if you have a chance. Thanks a lot!

bbejeck reviewed (approved):
Thanks for the follow-up @lianetm - LGTM modulo fixing the tests.

lianetm committed:
Merge branch 'trunk' into lm-callbacks-docs

lianetm committed:
Fix flaky test

lianetm committed:
Merge branch 'trunk' into lm-callbacks-docs

lianetm commented:
Thanks @bbejeck ! Fixed the flaky test. Now the 3 new tests pass consistently locally (tried up to ~50 and all ok, used to fail under 10 iterations before the fix). Will wait for the build and trigger a few to validate on CI too.
